Kotlin precompiled script plugins fail to compile when plugins declared in inner classes are on the classpath

<!--- 
Please follow the instructions below. 
We receive dozens of issues every week, so to stay productive, we will close issues that don't provide enough information. 

Please open Android-related issues on the Android Issue Tracker at https://source.android.com/source/report-bugs
Please open Gradle Native-related issues at https://github.com/gradle/gradle-native/issues
-->

Within a precompiled script plugin in `buildSrc`, the code generation for plugin accessors generates an invalid documentation comment.

I've experienced this issue when working with the https://github.com/diffplug/goomph plugin suite, due to several plugin IDs referring to inner classes implementing the `Plugin` interface -- the error below provides specifics.

### Expected Behavior
The KDoc comments for plugin accessors should not cause the `buildSrc` project to fail to compile.

### Current Behavior
The `:buildSrc:compilePluginsBlocks` task fails:

<details>
<summary>Log</summary>

```
> Task :buildSrc:compilePluginsBlocks FAILED
e: C:\Users\Zach\dev\gradle-script-plugin-codegen-bug\buildSrc\build\generated-sources\kotlin-dsl-external-plugin-spec-builders\kotlin\gradle\kotlin\dsl\plugins\_6146be3f585cc77316cb14c3baab0f9d\PluginSpecBuilders.kt:167:126: Closing bracket expected
e: C:\Users\Zach\dev\gradle-script-plugin-codegen-bug\buildSrc\build\generated-sources\kotlin-dsl-external-plugin-spec-builders\kotlin\gradle\kotlin\dsl\plugins\_6146be3f585cc77316cb14c3baab0f9d\PluginSpecBuilders.kt:175:132: Closing bracket expected
e: C:\Users\Zach\dev\gradle-script-plugin-codegen-bug\buildSrc\build\generated-sources\kotlin-dsl-external-plugin-spec-builders\kotlin\gradle\kotlin\dsl\plugins\_6146be3f585cc77316cb14c3baab0f9d\PluginSpecBuilders.kt:183:120: Closing bracket expected
e: C:\Users\Zach\dev\gradle-script-plugin-codegen-bug\buildSrc\build\generated-sources\kotlin-dsl-external-plugin-spec-builders\kotlin\gradle\kotlin\dsl\plugins\_6146be3f585cc77316cb14c3baab0f9d\PluginSpecBuilders.kt:191:118: Closing bracket expected
e: C:\Users\Zach\dev\gradle-script-plugin-codegen-bug\buildSrc\build\generated-sources\kotlin-dsl-external-plugin-spec-builders\kotlin\gradle\kotlin\dsl\plugins\_6146be3f585cc77316cb14c3baab0f9d\PluginSpecBuilders.kt:199:126: Closing bracket expected
e: C:\Users\Zach\dev\gradle-script-plugin-codegen-bug\buildSrc\build\generated-sources\kotlin-dsl-external-plugin-spec-builders\kotlin\gradle\kotlin\dsl\plugins\_6146be3f585cc77316cb14c3baab0f9d\PluginSpecBuilders.kt:207:120: Closing bracket expected
e: C:\Users\Zach\dev\gradle-script-plugin-codegen-bug\buildSrc\build\generated-sources\kotlin-dsl-external-plugin-spec-builders\kotlin\gradle\kotlin\dsl\plugins\_6146be3f585cc77316cb14c3baab0f9d\PluginSpecBuilders.kt:231:103: Closing bracket expected
e: C:\Users\Zach\dev\gradle-script-plugin-codegen-bug\buildSrc\build\generated-sources\kotlin-dsl-external-plugin-spec-builders\kotlin\gradle\kotlin\dsl\plugins\_6146be3f585cc77316cb14c3baab0f9d\PluginSpecBuilders.kt:255:112: Closing bracket expected
e: C:\Users\Zach\dev\gradle-script-plugin-codegen-bug\buildSrc\build\generated-sources\kotlin-dsl-external-plugin-spec-builders\kotlin\gradle\kotlin\dsl\plugins\_6146be3f585cc77316cb14c3baab0f9d\PluginSpecBuilders.kt:279:100: Closing bracket expected
e: C:\Users\Zach\dev\gradle-script-plugin-codegen-bug\buildSrc\build\generated-sources\kotlin-dsl-external-plugin-spec-builders\kotlin\gradle\kotlin\dsl\plugins\_6146be3f585cc77316cb14c3baab0f9d\PluginSpecBuilders.kt:303:108: Closing bracket expected

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':buildSrc:compilePluginsBlocks'.
> Script compilation errors:

    Line 167:  * The `com.diffplug.gradle.eclipse.buildproperties` plugin implemented by [com.diffplug.gradle.eclipse.BuildPropertiesPlugin$Legacy].

                   ^ Closing bracket expected

    Line 175:  * The `com.diffplug.gradle.eclipse.excludebuildfolder` plugin implemented by [com.diffplug.gradle.eclipse.ExcludeBuildFolderPlugin$Legacy].

                         ^ Closing bracket expected

    Line 183:  * The `com.diffplug.gradle.eclipse.mavencentral` plugin implemented by [com.diffplug.gradle.eclipse.MavenCentralPlugin$Legacy].

             ^ Closing bracket expected

    Line 191:  * The `com.diffplug.gradle.eclipse.projectdeps` plugin implemented by [com.diffplug.gradle.eclipse.ProjectDepsPlugin$Legacy].

           ^ Closing bracket expected

    Line 199:  * The `com.diffplug.gradle.eclipse.resourcefilters` plugin implemented by [com.diffplug.gradle.eclipse.ResourceFiltersPlugin$Legacy].

                   ^ Closing bracket expected

    Line 207:  * The `com.diffplug.gradle.equinoxlaunch` plugin implemented by [com.diffplug.gradle.eclipserunner.EquinoxLaunchPlugin$Legacy].

             ^ Closing bracket expected

    Line 231:  * The `com.diffplug.gradle.oomph.ide` plugin implemented by [com.diffplug.gradle.oomph.OomphIdePlugin$Legacy].
                                                                                                                    ^ Closing bracket expected

    Line 255:  * The `com.diffplug.gradle.osgi.bndmanifest` plugin implemented by [com.diffplug.gradle.osgi.BndManifestPlugin$Legacy].

     ^ Closing bracket expected

    Line 279:  * The `com.diffplug.gradle.p2.asmaven` plugin implemented by [com.diffplug.gradle.p2.AsMavenPlugin$Legacy].
                                                                                                                 ^ Closing bracket expected

    Line 303:  * The `com.diffplug.gradle.swt.nativedeps` plugin implemented by [com.diffplug.gradle.swt.NativeDepsPlugin$Legacy].

 ^ Closing bracket expected

  10 errors

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 27s
```
</details>

### Context
This prevents me from referring to any plugins in the `com.diffplug:goomph` module within my project's precompiled script plugins.

### Steps to Reproduce 

See the buildSrc project from https://github.com/zml2008/gradle-script-plugin-codegen-bug

### Your Environment
Reproducible on Gradle 6.8.3 and 7.1.1

--- a/subprojects/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/PrecompiledScriptPluginIntegrationTest.kt
+++ b/subprojects/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/PrecompiledScriptPluginIntegrationTest.kt
@@ -14,6 +14,7 @@ import org.hamcrest.MatcherAssert.assertThat
 import org.junit.Assert.assertFalse
 import org.junit.Assert.assertTrue
 import org.junit.Test
+import spock.lang.Issue
 
 
 @LeaksFileHandles("Kotlin Compiler Daemon working directory")
@@ -250,6 +251,7 @@ class PrecompiledScriptPluginIntegrationTest : AbstractPluginIntegrationTest() {
     }
 
     @Test
+    @Issue("https://github.com/gradle/gradle/issues/17619")
     @ToBeFixedForConfigurationCache(because = "Kotlin Gradle Plugin")
     fun `accessors are available after registering plugin`() {
         withSettings(
@@ -318,8 +320,11 @@ class PrecompiledScriptPluginIntegrationTest : AbstractPluginIntegrationTest() {
                     "src/main/java/producer/ProducerPlugin.java",
                     """
                     package producer;
-                    public class ProducerPlugin implements ${nameOf<Plugin<*>>()}<${nameOf<Project>()}> {
-                       @Override public void apply(${nameOf<Project>()} target) {}
+                    public class ProducerPlugin {
+                        // Using internal class to verify https://github.com/gradle/gradle/issues/17619
+                        public static class Implementation implements ${nameOf<Plugin<*>>()}<${nameOf<Project>()}> {
+                            @Override public void apply(${nameOf<Project>()} target) {}
+                        }
                     }
                     """
                 )
@@ -340,7 +345,7 @@ class PrecompiledScriptPluginIntegrationTest : AbstractPluginIntegrationTest() {
                     plugins {
                         producer {
                             id = 'producer-plugin'
-                            implementationClass = 'producer.ProducerPlugin'
+                            implementationClass = 'producer.ProducerPlugin${'$'}Implementation'
                         }
                     }
                 }

--- a/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/PluginAccessorsClassPath.kt
+++ b/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/PluginAccessorsClassPath.kt
@@ -40,6 +40,7 @@ import org.gradle.kotlin.dsl.codegen.fileHeader
 import org.gradle.kotlin.dsl.codegen.fileHeaderFor
 import org.gradle.kotlin.dsl.codegen.kotlinDslPackagePath
 import org.gradle.kotlin.dsl.codegen.pluginEntriesFrom
+import org.gradle.kotlin.dsl.codegen.sourceNameOfBinaryName
 import org.gradle.kotlin.dsl.concurrent.IO
 import org.gradle.kotlin.dsl.concurrent.withAsynchronousIO
 import org.gradle.kotlin.dsl.concurrent.withSynchronousIO
@@ -333,7 +334,7 @@ fun BufferedWriter.appendSourceCodeForPluginAccessors(
                     format(
                         """
                         /**
-                         * The `$id` plugin implemented by [${implementationClass.replace('$', '.')}].
+                         * The `$id` plugin implemented by [${sourceNameOfBinaryName(implementationClass)}].
                          */
                         val `$extendedType`.`${extension.name}`: PluginDependencySpec
                             get() = $pluginsRef.id("$id")

--- a/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/PluginAccessorsClassPath.kt
+++ b/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/accessors/PluginAccessorsClassPath.kt
@@ -333,7 +333,7 @@ fun BufferedWriter.appendSourceCodeForPluginAccessors(
                     format(
                         """
                         /**
-                         * The `$id` plugin implemented by [$implementationClass].
+                         * The `$id` plugin implemented by [${implementationClass.replace('$', '.')}].
                          */
                         val `$extendedType`.`${extension.name}`: PluginDependencySpec
                             get() = $pluginsRef.id("$id")

--- a/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/codegen/ApiTypeProvider.kt
+++ b/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/codegen/ApiTypeProvider.kt
@@ -567,7 +567,7 @@ fun binaryNameOfInternalName(internalName: String): String =
     Type.getObjectType(internalName).className
 
 
-private
+internal
 fun sourceNameOfBinaryName(binaryName: String): String =
     when (binaryName) {
         "void" -> "Unit"

